File: c4-4-dandc.js
function binarySearch(arr, search) {
  if (arr.length === 1) {
    if (arr[0] === search) {
      return arr[0];
    } else {
      return undefined;
    }
  }

  let low = 0;
  let high = arr.length;
  const mid = Math.floor((low + high) / 2);

  const guess = arr[mid];

  if (guess === search) {
    return guess;
  } else {
    if (guess > search) {
      high = mid;
    } else {
      low = mid;
    }
    return binarySearch(arr.slice(low, high), search);
  }
}

const myList = [1, 3, 5, 7, 9];

console.log(binarySearch(myList, 3));
console.log(binarySearch(myList, -1));

File: python/c4-4-dandc.py
def binary_search(arr, search):
    if len(arr) == 1:
        if arr[0] == search:
            return arr[0]
        else:
            return None
    low = 0
    high = len(arr)
    mid = (low + high) // 2

    guess = arr[mid]

    if guess == search:
        return guess
    else:
        if guess > search:
            high = mid
        else:
            low = mid

        return binary_search(arr[low:high], search)


my_list = [1, 3, 5, 7, 9]

print(binary_search(my_list, 3))
print(binary_search(my_list, -1))
